### DISCUSSION TITLE
RAM小サイズ向けのサンプル

### DISCUSSION THREAD

**Discussion by Y-Ishioka** [toppers/users-forum#13] [General]
asp-1.9.3.tar.gz とasp_arch_arm_m0_gcc-1.9.4.tar.gz を使って target/stm32l073nucleo64_gcc をコンフィギュレーションしてビルドしたんですけど、RAMが20KBしかないのでsampleのビルドに失敗します。
RAM小サイズ向けのサンプルってないでしょうか。
あと、可能なら他の人もTOPPERSカーネルをダウンロードしてビルドの確認ができるよう sample2とかみたいにカーネルソースに同梱されてると便利で助かります。

**Comment by hiroakitakada** [comment]
RAMを最も多く使っているのが，スタック領域だと思います。デフォルトでは，各タスクのスタック領域を多めに（具体的には，4KB）取っているため，これを減らしてもらえれば20KBに収まるように思います。具体的には，target_test.hでSTACK_SIZEを小さい値（例えば，1KB）に定義してもらえれば良いと思います。
ただし，システムログタスクのスタックサイズだけは，STACK_SIZEでは変更できません。これを変更するには，target.cdl中のDefaultTaskStackSizeの値を変更することが必要です。
RAMが20KBですと，スタックサイズをまじめに考えて設定する必要があるように思います。

**Reply by Y-Ishioka** [reply]
アドバイスありがとうございます。
サンプルのスタックが4096バイトで大きめにとってますね。3072バイト（3KB）にしたらギリギリ20KBを下回りました。
STM32L072-Nucleoは持っていないのでCortex-M4ソースを使って定義を変更してF401用でビルドし、STM32F401-Nucleoで動作確認したところsample1.cのタスクが動作するのを確認できました。
STACK_SIZEを3072にしておくとRAM 20KBに対応できそうです。

vi ../../asp/target/stm32f401nucleo_gcc/stm32f4xx_rom.ld
ーー
MEMORY
{
    ROM (rx) : ORIGIN = 0x08000000, LENGTH = 1024K
    RAM (rwx) : ORIGIN = 0x20000000, LENGTH = 20K
}

vi sample1.h
ーー
#define STACK_SIZE              3072

ーー
20004eb0 B _kernel_current_time
20004eb4 B _kernel_next_time
20004eb8 B _kernel_last_index
20004ebc B __bss_end
20004ebc B _end
